---
title: "X and W in Allen data"
author: "Fanny Perraudeau"
date: "10/23/2016"
output: 
  html_document: 
    fig_height: 10
    fig_width: 10
    toc: yes
    code_folding: hide
    toc_float: yes
---

```{r options, echo=FALSE, results="hide",mesasge=FALSE, error=FALSE, include=FALSE, autodep=TRUE}
knitr::opts_chunk$set(fig.align="center", cache=TRUE, error=FALSE, message=FALSE, warning=TRUE)
library(SummarizedExperiment)
library(scRNAseq)
library(zinb)
library(scran)
library(EDASeq)
library(ggplot2)
library(magrittr)
library(matrixStats)
if (!require(digest)) install.packages('digest')

library(RColorBrewer)
cols <- brewer.pal(8, "Set1")
cols2 <- c(brewer.pal(8, "Set2"), brewer.pal(8, "Set3"), brewer.pal(8, "Set1"))

set.seed(9938)
```

# TL;DR

We add cell type information in X and look at W.

# Data

As a first pass, I will only focus on the cells that passed the QC filters (by the original authors) and that were part of the "core" clusters. I will color-code the cells by either known cell type or by inferred cluster (inferred in the original study).

Select cells, remove ERCC spike-ins, filter out the genes that do not have at least 10 counts in at least 10 cells.

```{r datain}
data("allen")
allen_core <- allen[grep("^ERCC-", rownames(allen), invert = TRUE),
                    which(colData(allen)$Core.Type=="Core")]

filter <- apply(assay(allen_core)>10, 1, sum)>=10
```

Number of retained genes:
```{r}
print(sum(filter))
```

To speed up the computations, I will focus on the top 1,000 most variable genes.

```{r filtering}
allen_core <- allen_core[filter,]
core <- assay(allen_core)

vars <- rowVars(log1p(core))
names(vars) <- rownames(core)
vars <- sort(vars, decreasing = TRUE)
core <- core[names(vars)[1:1000],]
```

First, let's look at PCA (of the log counts) for reference.

```{r pca}
par(mfrow=c(1, 2))
bio <- as.factor(colData(allen_core)$driver_1_s)
cl <- as.factor(colData(allen_core)$Primary.Type)

detection_rate <- colSums(core>0)
coverage <- colSums(core)

pca <- prcomp(t(log1p(core)))
plot(pca$x, col=cols[bio], pch=19, main="PCA of log-counts, centered not scaled")
legend("bottomleft", levels(bio), fill=cols)
plot(pca$x, col=cols2[cl], pch=19, main="PCA of log-counts, centered not scaled")

```

```{r pca_cor}
df <- data.frame(PC1=pca$x[,1], PC2=pca$x[,2], detection_rate=detection_rate, coverage=coverage)
pairs(df, pch=19, col=cols[bio])

print(cor(df, method="spearman"))
```

# X cell types, W
Both Vmu and Vpi have intercept, common dispersion is TRUE.

```{r zinb}
print(system.time(zinb_X <- zinbFit(core, ncores = 3, K = 2,
                                    X=model.matrix(~bio))))
```

Plot the results with cells colored according to their biological condition.

```{r zinb_plot}
plot(zinb_X@W, col=cols[bio], pch=19, xlab="W1", ylab="W2", main="Both V intercepts")
legend("bottomright", levels(bio), fill=cols)

plot(zinb_X@W, col=cols2[cl], pch=19, xlab="W1", ylab="W2", main="Both V intercepts")
legend("topleft", levels(cl), fill=cols2)
```

## Explore Gamma estimates

One interpretation of the model is that $\gamma_mu$ will act as a "size factor". Here, we check whether it captures sequencing depth and/or detection rate.

```{r gamma}
#total number of detected genes in the cell
df <- data.frame(W1=zinb_X@W[,1],
                 W2=zinb_X@W[,2],
                 gamma_mu = zinb_X@gamma_mu[1,],
                 gamma_pi = zinb_X@gamma_pi[1,], 
                 detection_rate=detection_rate, 
                 coverage=coverage)
pairs(df, pch=19, col=cols[bio])
print(cor(df, method="spearman"))
```

$\gamma_pi$ clearly captures the detection rate and $\gamma_mu$ clearly captures the coverage.
Interestingly, the two estimates are positively correlated.

```{r mu_pi}
par(mfrow=c(1, 2))
plot(rowMeans(getPi(zinb_X)), detection_rate, xlab="Average estimated Pi", ylab="detection rate", pch=19, col=cols[bio])

plot(rowMeans(log1p(getMu(zinb_X))), coverage, xlab="Average estimated log Mu", ylab="coverage", pch=19, col=cols[bio])
```

It is worth looking at the other estimates as well, namely, $\beta$ and $\alpha$.

```{r other_pars}
par(mfrow=c(3, 2))
boxplot(zinb_X@beta_mu[1,], main="Beta_Mu")
abline(h=0)
boxplot(zinb_X@beta_pi[1,], main="Beta_Pi")
abline(h=0)
boxplot(zinb_X@beta_mu[2,], main="Beta_Mu")
abline(h=0)
boxplot(zinb_X@beta_pi[2,], main="Beta_Pi")
abline(h=0)
boxplot(zinb_X@beta_mu[3,], main="Beta_Mu")
abline(h=0)
boxplot(zinb_X@beta_pi[3,], main="Beta_Pi")
abline(h=0)

par(mfrow=c(2, 2))
plot(t(zinb_X@alpha_mu), xlab="Alpha_Mu 1", ylab="Alpha_Mu 2", main="Alpha_Mu")
plot(t(zinb_X@alpha_pi), xlab="Alpha_Pi 1", ylab="Alpha_Pi 2", main="Alpha_Pi")
```

## Highest beta and alpha

beta_1 : Rbp4 VS (layer5) Ntsr1 (layer6)  
beta_2 : Scnn1a (layer4) VS Ntsr1 (layer6)

```{r de}
get_DE <- function(zinb_res, core, beta = T, mu = T, component = 2, n=10){
  if (beta & mu){
    mat = zinb_X@beta_mu[component, ]
  }else if (beta & !mu){
    mat = zinb_X@beta_pi[component, ]
  }else if (!beta & mu){
    mat = zinb_X@alpha_mu[component, ]
  }else{
    mat = zinb_X@alpha_pi[component, ]
  }
  idx = order(-abs(mat))[1:n]
  de = mat[idx]
  names(de) = rownames(core)[idx]
  de
}
#beta_mu
get_DE(zinb_res, core, beta = T, mu = T, 2)
get_DE(zinb_res, core, beta = T, mu = T, 3)
#beta_pi
get_DE(zinb_res, core, beta = T, mu = F, 2)
get_DE(zinb_res, core, beta = T, mu = F, 3)
#alpha_mu
get_DE(zinb_res, core, beta = F, mu = T, 1)
get_DE(zinb_res, core, beta = F, mu = T, 2)
#alpha_pi
get_DE(zinb_res, core, beta = F, mu = F, 1)
get_DE(zinb_res, core, beta = F, mu = F, 2)
```

### All 1,000 most variable genes, before fit
```{r heatmap_all}
require(gplots)
# all 1,000 most variable genes, before fit
heatmap.2(log1p(core), ColSideColor = cols[bio], trace = 'none')
```

### All 1,000 most variable genes, after fit
```{r heatmaps_fit}
heatmap.2(log1p(t(getMu(zinb_X))), ColSideColor = cols[bio], trace = 'none')
heatmap.2(log1p(t(getMu(zinb_X))), ColSideColor = cols2[cl], trace = 'none')

heatmap.2(t(getPi(zinb_X)), ColSideColor = cols[bio], trace = 'none')
heatmap.2(t(getPi(zinb_X)), ColSideColor = cols2[cl], trace = 'none')
```

### highest beta_mu, after fit
```{r heatmaps_beta_mu}
rbp4_ntsr1 = names(get_DE(zinb_res, core, beta = T, mu = T, 2,n=25))
scnn1a_ntsr1 = names(get_DE(zinb_res, core, beta = T, mu = T, 3, n=25))
mu_de = unique(c(rbp4_ntsr1, scnn1a_ntsr1))
length(mu_de)
fitted_mu = log1p(t(getMu(zinb_X)))
rownames(fitted_mu) = rownames(core)
fitted_mu = fitted_mu[mu_de, ]
heatmap.2(fitted_mu, ColSideColor = cols[bio], trace = 'none')
heatmap.2(fitted_mu, ColSideColor = cols2[cl], trace = 'none')
```

### highest beta_pi, after fit
```{r heatmaps_beta_pi}
rbp4_ntsr1 = names(get_DE(zinb_res, core, beta = T, mu = F, 2,n=50))
scnn1a_ntsr1 = names(get_DE(zinb_res, core, beta = T, mu = F, 3, n=50))
pi_de = unique(c(rbp4_ntsr1, scnn1a_ntsr1))
length(pi_de)
fitted_pi = t(getPi(zinb_X))
rownames(fitted_pi) = rownames(core)
fitted_pi = fitted_pi[pi_de, ]
heatmap.2(fitted_pi, ColSideColor = cols[bio], trace = 'none')
heatmap.2(fitted_pi, ColSideColor = cols2[cl], trace = 'none')

```

### highest alpha_mu, after fit
```{r heatmaps_fit_alpha_mu}
rbp4_ntsr1 = names(get_DE(zinb_res, core, beta = F, mu = T, 1,n=25))
scnn1a_ntsr1 = names(get_DE(zinb_res, core, beta = F, mu = T, 2, n=25))
mu_de = unique(c(rbp4_ntsr1, scnn1a_ntsr1))
length(mu_de)
fitted_mu = log1p(t(getMu(zinb_X)))
rownames(fitted_mu) = rownames(core)
fitted_mu = fitted_mu[mu_de, ]
heatmap.2(fitted_mu, ColSideColor = cols[bio], trace = 'none')
heatmap.2(fitted_mu, ColSideColor = cols2[cl], trace = 'none')

```

### highest alpha_pi, after fit
```{r heatmaps_fit_alpha_pi}
rbp4_ntsr1 = names(get_DE(zinb_res, core, beta = F, mu = F, 1,n=25))
scnn1a_ntsr1 = names(get_DE(zinb_res, core, beta = F, mu = F, 2, n=25))
pi_de = unique(c(rbp4_ntsr1, scnn1a_ntsr1))
length(pi_de)
fitted_pi = t(getPi(zinb_X))
rownames(fitted_pi) = rownames(core)
fitted_pi = fitted_pi[pi_de, ]
heatmap.2(fitted_pi, ColSideColor = cols[bio], trace = 'none')
heatmap.2(fitted_pi, ColSideColor = cols2[cl], trace = 'none')
```

## Goodness of fit
```{r zinb_MD}
myPlotMD <- function(x, y, main= '', xlab = 'Mean', ylab = 'Difference'){
  mean = .5*(x + y)
  diff = x - y
  smoothScatter(mean, diff, xlab = xlab, ylab = ylab, main = main)
  abline(h = 0, col = 'red', lwd = 2)
}

myPlotMD(log2(rowMeans(t(getMu(zinb_X)))), log2(rowMeans(core)),
         main = 'estimated vs observed mean count')

estimated = unlist(t(getPi(zinb_X)))
observed = rep(rowMeans(core > 0), each = ncol(core))
myPlotMD(estimated, observed, main = 'estimated vs observed zero probability')

mu = t(getMu(zinb_X))
mu6 = mu[, as.vector(bio) == 'Ntsr1-Cre_GN220']
mu5 = mu[, as.vector(bio) == 'Rbp4-Cre_KL100']
mu6 = log2(rowMeans(mu6))
mu5 = log2(rowMeans(mu5))
myPlotMD(mu6, mu5, main="MD L6 vs L5, estimated mu")
muDE = names(get_DE(zinb_res, core, beta = T, mu = T, 2, n = 25))
muIdx = rownames(core) %in% muDE
points(.5*(mu6 + mu5)[muIdx], (mu6 - mu5)[muIdx], col="cyan", pch=1, lwd=2)
piDE = names(get_DE(zinb_res, core, beta = T, mu = F, 2, n = 25))
piIdx = rownames(core) %in% piDE
points(.5*(mu6 + mu5)[piIdx], (mu6 - mu5)[piIdx], col="magenta", pch=3, lwd=2)

pi = t(getPi(zinb_X))
pi6 = pi[, as.vector(bio) == 'Ntsr1-Cre_GN220']
pi5 = pi[, as.vector(bio) == 'Rbp4-Cre_KL100']
pi6 = rowMeans(pi6)
pi5 = rowMeans(pi5)
myPlotMD(pi6, pi5, main = "MD L5 vs L4, estimated pi")
points(.5*(pi6 + pi5)[muIdx], (pi6 - pi5)[muIdx], col="cyan", pch=1, lwd=2)
points(.5*(pi6 + pi5)[piIdx], (pi6 - pi5)[piIdx], col="magenta", pch=3, lwd=2)

```

# X cell types, no W
Both Vmu and Vpi have intercept, common dispersion is TRUE.

```{r zinb_noW}
print(system.time(zinb_X <- zinbFit(core, ncores = 3, K = 0,
                                    X=model.matrix(~bio))))
```

```{r mu_pi_noW}
par(mfrow=c(1, 2))
plot(rowMeans(getPi(zinb_X)), detection_rate, xlab="Average estimated Pi", ylab="detection rate", pch=19, col=cols[bio])

plot(rowMeans(log1p(getMu(zinb_X))), coverage, xlab="Average estimated log Mu", ylab="coverage", pch=19, col=cols[bio])
```

It is worth looking at the other estimates as well, namely, $\beta$ and $\alpha$.

```{r other_pars_noW}
par(mfrow=c(3, 2))
boxplot(zinb_X@beta_mu[1,], main="Beta_Mu")
abline(h=0)
boxplot(zinb_X@beta_pi[1,], main="Beta_Pi")
abline(h=0)
boxplot(zinb_X@beta_mu[2,], main="Beta_Mu")
abline(h=0)
boxplot(zinb_X@beta_pi[2,], main="Beta_Pi")
abline(h=0)
boxplot(zinb_X@beta_mu[3,], main="Beta_Mu")
abline(h=0)
boxplot(zinb_X@beta_pi[3,], main="Beta_Pi")
abline(h=0)
```

## Highest beta and alpha

beta_1 : Rbp4 (layer5) VS Ntsr1 (layer6)  
beta_2 : Scnn1a (layer4) VS Ntsr1 (layer6)

```{r de_noW}
#beta_mu
get_DE(zinb_res, core, beta = T, mu = T, 2)
get_DE(zinb_res, core, beta = T, mu = T, 3)
#beta_pi
get_DE(zinb_res, core, beta = T, mu = F, 2)
get_DE(zinb_res, core, beta = T, mu = F, 3)
```

### All 1,000 most variable genes, after fit
```{r heatmaps_fit_noW}
heatmap.2(log1p(t(getMu(zinb_X))), ColSideColor = cols[bio], trace = 'none')
heatmap.2(log1p(t(getMu(zinb_X))), ColSideColor = cols2[cl], trace = 'none')

heatmap.2(t(getPi(zinb_X)), ColSideColor = cols[bio], trace = 'none')
heatmap.2(t(getPi(zinb_X)), ColSideColor = cols2[cl], trace = 'none')
```

### highest beta_mu, after fit
```{r heatmaps_beta_mu_noW}
rbp4_ntsr1 = names(get_DE(zinb_res, core, beta = T, mu = T, 2,n=25))
scnn1a_ntsr1 = names(get_DE(zinb_res, core, beta = T, mu = T, 3, n=25))
mu_de = unique(c(rbp4_ntsr1, scnn1a_ntsr1))
length(mu_de)
fitted_mu = log1p(t(getMu(zinb_X)))
rownames(fitted_mu) = rownames(core)
fitted_mu = fitted_mu[mu_de, ]
heatmap.2(fitted_mu, ColSideColor = cols[bio], trace = 'none')
heatmap.2(fitted_mu, ColSideColor = cols2[cl], trace = 'none')
```

### highest beta_pi, after fit
```{r heatmaps_beta_pi_noW}
rbp4_ntsr1 = names(get_DE(zinb_res, core, beta = T, mu = F, 2,n=25))
scnn1a_ntsr1 = names(get_DE(zinb_res, core, beta = T, mu = F, 3, n=25))
pi_de = unique(c(rbp4_ntsr1, scnn1a_ntsr1))
length(pi_de)
fitted_pi = t(getPi(zinb_X))
rownames(fitted_pi) = rownames(core)
fitted_pi = fitted_pi[pi_de, ]
heatmap.2(fitted_pi, ColSideColor = cols[bio], trace = 'none')
heatmap.2(fitted_pi, ColSideColor = cols2[cl], trace = 'none')

```

## Goodness of fit
```{r zinb_MD_noW}
myPlotMD(log2(rowMeans(t(getMu(zinb_X)))), log2(rowMeans(core)),
         main = 'estimated vs observed mean count')

estimated = unlist(t(getPi(zinb_X)))
observed = rep(rowMeans(core > 0), each = ncol(core))
myPlotMD(estimated, observed, main = 'estimated vs observed zero probability')

mu = t(getMu(zinb_X))
mu6 = mu[, as.vector(bio) == 'Ntsr1-Cre_GN220']
mu5 = mu[, as.vector(bio) == 'Rbp4-Cre_KL100']
mu6 = log2(rowMeans(mu6))
mu5 = log2(rowMeans(mu5))
myPlotMD(mu6, mu5, main="MD L6 vs L5, estimated mu")
muDE = names(get_DE(zinb_res, core, beta = T, mu = T, 2, n = 25))
muIdx = rownames(core) %in% muDE
points(.5*(mu6 + mu5)[muIdx], (mu6 - mu5)[muIdx], col="cyan", pch=1, lwd=2)
piDE = names(get_DE(zinb_res, core, beta = T, mu = F, 2, n = 25))
piIdx = rownames(core) %in% piDE
points(.5*(mu6 + mu5)[piIdx], (mu6 - mu5)[piIdx], col="magenta", pch=3, lwd=2)

pi = t(getPi(zinb_X))
pi6 = pi[, as.vector(bio) == 'Ntsr1-Cre_GN220']
pi5 = pi[, as.vector(bio) == 'Rbp4-Cre_KL100']
pi6 = rowMeans(pi6)
pi5 = rowMeans(pi5)
myPlotMD(pi6, pi5, main = "MD L5 vs L4, estimated pi")
points(.5*(pi6 + pi5)[muIdx], (pi6 - pi5)[muIdx], col="cyan", pch=1, lwd=2)
points(.5*(pi6 + pi5)[piIdx], (pi6 - pi5)[piIdx], col="magenta", pch=3, lwd=2)

```








