---
title: "X and W in Allen data"
author: "Fanny Perraudeau"
date: "10/23/2016"
output: 
  html_document: 
    fig_height: 10
    fig_width: 10
    toc: yes
    code_folding: hide
    toc_float: yes
---

```{r options, echo=FALSE, results="hide",mesasge=FALSE, error=FALSE, include=FALSE, autodep=TRUE}
knitr::opts_chunk$set(fig.align="center", cache=TRUE, error=FALSE, message=FALSE, warning=TRUE)
library(SummarizedExperiment)
library(scRNAseq)
library(zinb)
library(scran)
library(EDASeq)
library(ggplot2)
library(magrittr)
library(matrixStats)
if (!require(digest)) install.packages('digest')

library(RColorBrewer)
cols <- brewer.pal(8, "Set1")
cols2 <- c(brewer.pal(8, "Set2"), brewer.pal(8, "Set3"), brewer.pal(8, "Set1"))

set.seed(9938)
```

## TL;DR

We add cell type information in X and look at W.

## Data

As a first pass, I will only focus on the cells that passed the QC filters (by the original authors) and that were part of the "core" clusters. I will color-code the cells by either known cell type or by inferred cluster (inferred in the original study).

Select cells, remove ERCC spike-ins, filter out the genes that do not have at least 10 counts in at least 10 cells.

```{r datain}
data("allen")
allen_core <- allen[grep("^ERCC-", rownames(allen), invert = TRUE),
                    which(colData(allen)$Core.Type=="Core")]

filter <- apply(assay(allen_core)>10, 1, sum)>=10
```

Number of retained genes:
```{r}
print(sum(filter))
```

To speed up the computations, I will focus on the top 1,000 most variable genes.

```{r filtering}
allen_core <- allen_core[filter,]
core <- assay(allen_core)

vars <- rowVars(log1p(core))
names(vars) <- rownames(core)
vars <- sort(vars, decreasing = TRUE)
core <- core[names(vars)[1:1000],]
```

First, let's look at PCA (of the log counts) for reference.

```{r pca}
par(mfrow=c(1, 2))
bio <- as.factor(colData(allen_core)$driver_1_s)
cl <- as.factor(colData(allen_core)$Primary.Type)

detection_rate <- colSums(core>0)
coverage <- colSums(core)

pca <- prcomp(t(log1p(core)))
plot(pca$x, col=cols[bio], pch=19, main="PCA of log-counts, centered not scaled")
legend("bottomleft", levels(bio), fill=cols)
plot(pca$x, col=cols2[cl], pch=19, main="PCA of log-counts, centered not scaled")

```

```{r pca_cor}
df <- data.frame(PC1=pca$x[,1], PC2=pca$x[,2], detection_rate=detection_rate, coverage=coverage)
pairs(df, pch=19, col=cols[bio])

print(cor(df, method="spearman"))
```

## X with cell types, common disp

```{r zinb}
print(system.time(zinb_X <- zinbFit(core, ncores = 3, K = 2,
                                    X=model.matrix(~bio))))
```

Plot the results with cells colored according to their biological condition.

```{r zinb_plot}
par(mfrow=c(1, 2))
plot(zinb_X@W, col=cols[bio], pch=19, xlab="W1", ylab="W2", main="Both V intercepts")
legend("bottomright", levels(bio), fill=cols)

plot(zinb_X@W, col=cols2[cl], pch=19, xlab="W1", ylab="W2", main="Both V intercepts")
legend("bottom", levels(cl), fill=cols2, horiz=T)
```

### Explore Gamma estimates

One interpretation of the model is that $\gamma_mu$ will act as a "size factor". Here, we check whether it captures sequencing depth and/or detection rate.

```{r gamma}
#total number of detected genes in the cell
df <- data.frame(W1=zinb_X@W[,1], W2=zinb_X@W[,2], gamma_mu = zinb_X@gamma_mu[1,], gamma_pi = zinb_X@gamma_pi[1,], detection_rate=detection_rate, coverage=coverage)
pairs(df, pch=19, col=cols[bio])

print(cor(df, method="spearman"))
```

$\gamma_pi$ clearly captures the detection rate and $\gamma_mu$ clearly captures the coverage.
Interestingly, the two estimates are positively correlated.

```{r mu_pi}
par(mfrow=c(1, 2))
plot(rowMeans(getPi(zinb_X)), detection_rate, xlab="Average estimated Pi", ylab="detection rate", pch=19, col=cols[bio])

plot(rowMeans(log1p(getMu(zinb_X))), coverage, xlab="Average estimated log Mu", ylab="coverage", pch=19, col=cols[bio])
```

It is worth looking at the other estimates as well, namely, $\beta$ and $\alpha$.

```{r other_pars}
par(mfrow=c(3, 2))
boxplot(zinb_X@beta_mu[1,], main="Beta_Mu")
abline(h=0)
boxplot(zinb_X@beta_pi[1,], main="Beta_Pi")
abline(h=0)
boxplot(zinb_X@beta_mu[2,], main="Beta_Mu")
abline(h=0)
boxplot(zinb_X@beta_pi[2,], main="Beta_Pi")
abline(h=0)
boxplot(zinb_X@beta_mu[3,], main="Beta_Mu")
abline(h=0)
boxplot(zinb_X@beta_pi[3,], main="Beta_Pi")
abline(h=0)

par(mfrow=c(2, 2))
plot(t(zinb_X@alpha_mu), xlab="Alpha_Mu 1", ylab="Alpha_Mu 2", main="Alpha_Mu")
plot(t(zinb_X@alpha_pi), xlab="Alpha_Pi 1", ylab="Alpha_Pi 2", main="Alpha_Pi")
```


## X with cell types, diff disp

```{r zinb_disp}
print(system.time(zinb_Xdisp <- zinbFit(core, ncores = 3, K = 2,
                                    X=model.matrix(~bio),commondispersion=F)))
```

Plot the results with cells colored according to their biological condition.

```{r zinb_plot_disp}
par(mfrow=c(1, 2))
plot(zinb_Xdisp@W, col=cols[bio], pch=19, xlab="W1", ylab="W2", main="Both V intercepts")
legend("bottomright", levels(bio), fill=cols)

plot(zinb_Xdisp@W, col=cols2[cl], pch=19, xlab="W1", ylab="W2", main="Both V intercepts")
legend("bottom", levels(cl), fill=cols2,horiz = T)
```

### Explore Gamma estimates

One interpretation of the model is that $\gamma_mu$ will act as a "size factor". Here, we check whether it captures sequencing depth and/or detection rate.

```{r gamma_disp}
#total number of detected genes in the cell
df <- data.frame(W1=zinb_Xdisp@W[,1], W2=zinb_Xdisp@W[,2], gamma_mu = zinb_Xdisp@gamma_mu[1,], gamma_pi = zinb_Xdisp@gamma_pi[1,], detection_rate=detection_rate, coverage=coverage)
pairs(df, pch=19, col=cols[bio])

print(cor(df, method="spearman"))
```

$\gamma_pi$ clearly captures the detection rate and $\gamma_mu$ clearly captures the coverage.
Interestingly, the two estimates are positively correlated.

```{r mu_pi_disp}
par(mfrow=c(1, 2))
plot(rowMeans(getPi(zinb_Xdisp)), detection_rate, xlab="Average estimated Pi", ylab="detection rate", pch=19, col=cols[bio])

plot(rowMeans(log1p(getMu(zinb_Xdisp))), coverage, xlab="Average estimated log Mu", ylab="coverage", pch=19, col=cols[bio])
```

It is worth looking at the other estimates as well, namely, $\beta$ and $\alpha$.

```{r other_pars_disp}
par(mfrow=c(3, 2))
boxplot(zinb_Xdisp@beta_mu[1,], main="Beta_Mu")
abline(h=0)
boxplot(zinb_Xdisp@beta_pi[1,], main="Beta_Pi")
abline(h=0)
boxplot(zinb_Xdisp@beta_mu[2,], main="Beta_Mu")
abline(h=0)
boxplot(zinb_Xdisp@beta_pi[2,], main="Beta_Pi")
abline(h=0)
boxplot(zinb_Xdisp@beta_mu[3,], main="Beta_Mu")
abline(h=0)
boxplot(zinb_Xdisp@beta_pi[3,], main="Beta_Pi")
abline(h=0)

par(mfrow=c(1, 2))
plot(t(zinb_Xdisp@alpha_mu), xlab="Alpha_Mu 1", ylab="Alpha_Mu 2", main="Alpha_Mu")
plot(t(zinb_Xdisp@alpha_pi), xlab="Alpha_Pi 1", ylab="Alpha_Pi 2", main="Alpha_Pi")
```


## X with cell types, no V intercept

```{r zinb_noInt}
print(system.time(zinb_XnoneV <- zinbFit(core, ncores = 3, K = 2,
                                    X=model.matrix(~bio),
                                    V=matrix(0, ncol=1, nrow=NROW(core)))))
```

Plot the results with cells colored according to their biological condition.

```{r zinb_plot_noInt}
par(mfrow=c(1, 2))
plot(zinb_XnoneV@W, col=cols[bio], pch=19, xlab="W1", ylab="W2", main="Both V intercepts")
legend("bottomright", levels(bio), fill=cols)

plot(zinb_XnoneV@W, col=cols2[cl], pch=19, xlab="W1", ylab="W2", main="Both V intercepts")
legend("bottom", levels(cl), fill=cols2, horiz=T)
```

### Explore Gamma estimates

One interpretation of the model is that $\gamma_mu$ will act as a "size factor". Here, we check whether it captures sequencing depth and/or detection rate.

```{r gamma_noInt}
#total number of detected genes in the cell
df <- data.frame(W1=zinb_XnoneV@W[,1], W2=zinb_XnoneV@W[,2], detection_rate=detection_rate, coverage=coverage)
pairs(df, pch=19, col=cols[bio])

print(cor(df, method="spearman"))
```

$\gamma_pi$ clearly captures the detection rate and $\gamma_mu$ clearly captures the coverage.
Interestingly, the two estimates are positively correlated.

```{r mu_pi_noInt}
par(mfrow=c(1, 2))
plot(rowMeans(getPi(zinb_XnoneV)), detection_rate, xlab="Average estimated Pi", ylab="detection rate", pch=19, col=cols[bio])

plot(rowMeans(log1p(getMu(zinb_XnoneV))), coverage, xlab="Average estimated log Mu", ylab="coverage", pch=19, col=cols[bio])
```

It is worth looking at the other estimates as well, namely, $\beta$ and $\alpha$.

```{r other_pars_noInt}
par(mfrow=c(3, 2))
boxplot(zinb_XnoneV@beta_mu[1,], main="Beta_Mu")
abline(h=0)
boxplot(zinb_XnoneV@beta_pi[1,], main="Beta_Pi")
abline(h=0)
boxplot(zinb_XnoneV@beta_mu[2,], main="Beta_Mu")
abline(h=0)
boxplot(zinb_XnoneV@beta_pi[2,], main="Beta_Pi")
abline(h=0)
boxplot(zinb_XnoneV@beta_mu[3,], main="Beta_Mu")
abline(h=0)
boxplot(zinb_XnoneV@beta_pi[3,], main="Beta_Pi")
abline(h=0)

par(mfrow=c(2, 2))
plot(t(zinb_XnoneV@alpha_mu), xlab="Alpha_Mu 1", ylab="Alpha_Mu 2", main="Alpha_Mu")
plot(t(zinb_XnoneV@alpha_pi), xlab="Alpha_Pi 1", ylab="Alpha_Pi 2", main="Alpha_Pi")
```

