---
title: 'Patel: covariates and batch effects'
author: "Davide Risso"
date: "2/23/2017"
output: 
  html_document: 
    fig_height: 7
    fig_width: 7
    toc: yes
    code_folding: hide
    toc_float: yes
---

```{r options, echo=FALSE, results="hide",mesasge=FALSE, error=FALSE, include=FALSE, autodep=TRUE}
knitr::opts_chunk$set(fig.align="center", cache=TRUE, error=FALSE, message=FALSE, warning=TRUE)
library(SummarizedExperiment)
library(scRNAseq)
library(zinbwave)
library(ggplot2)
library(magrittr)
library(matrixStats)
library(Rtsne)
library(scone)
library(RColorBrewer)
library(cluster)
library(parallel)
library(EDASeq)
library(digest)
library(edgeR)
library(DESeq2)

set.seed(8283)

library(BiocParallel)
library(doParallel)
registerDoParallel(6)
register(DoparParam())
```

Here, I will explore the relation between the expression values and some covariates provided with the data.

```{r datain}
counts <- read.table("patel/glioblastoma_raw_rnaseq_SCandbulk_counts_withannots.txt", header=TRUE, stringsAsFactors = FALSE)

info <- as.matrix(counts)[1,-(1:3)]
counts <- counts[-1,-1]

gene_symbols <- counts[,1]
ensembl_ids <- counts[,2]
sample_names <- colnames(counts)[-(1:2)]

all.counts <- counts[,-(1:2)]
rownames(all.counts) <- ensembl_ids

metadata <- read.table("patel/SraRunTable.txt", sep='\t', stringsAsFactors = FALSE, header=TRUE, row.names=5, na.strings = "<not provided>")
metadata <- metadata[sample_names,]

batch <- stringr::str_split_fixed(info, "_", 2)[,2]
batch <- stringr::str_split_fixed(batch, "_", 2)[,1]

# select only single-cell samples from patients
keep <- which(grepl("^Single cell", metadata$source_name_s) &
                !is.na(metadata$patient_id_s) &
                !is.na(metadata$subtype_s))
metadata <- metadata[keep,]
batch <- as.factor(batch[keep])
all.counts <- all.counts[,keep]
all.counts <- as.matrix(all.counts)
class(all.counts) <- "numeric"

stopifnot(all(rownames(metadata)==colnames(all.counts)))

col1 <- brewer.pal(9, "Set1")
col2 <- c(brewer.pal(8, "Set2"), brewer.pal(8, "Set3"), brewer.pal(8, "Set1"))

detection_rate <- colSums(all.counts>0)
coverage <- colSums(all.counts)

qc <- cbind(detection_rate, coverage)

level1 <- as.factor(metadata$patient_id_s[!is.na(metadata$subtype_s)])
level2 <- as.factor(metadata$subtype_s[!is.na(metadata$subtype_s)])
```

# Comparison of methods

```{r normalization}
filter <- rowSums(all.counts>50)>=50
raw <- all.counts[filter,]

totalcount = function (ei)
{
  sums = colSums(ei)
  eo = t(t(ei)*mean(sums)/sums)
  return(eo)
}

tc <- totalcount(raw)
fq <- FQT_FN(raw)
tmm <- TMM_FN(raw)
deseq <- DESEQ_FN(raw)
```

## PCA

```{r pca}
library(rARPACK)
fastpca <- function(expr, scale=FALSE) {
  svd_raw <- svds(scale(t(expr), center=TRUE, scale=scale), k=3, nu=3, nv=0)
  pc_raw <- svd_raw$u %*% diag(svd_raw$d[1:3])
  return(pc_raw)
}

vars <- rowVars(log1p(tc))
names(vars) <- rownames(tc)
vars <- sort(vars, decreasing = TRUE)
vargenes <- names(vars)[1:1000]

pc_raw <- fastpca(log1p(raw[vargenes,]))
pc_tc <- fastpca(log1p(tc[vargenes,]))
pc_fq <- fastpca(log1p(fq[vargenes,]))
pc_tmm <- fastpca(log1p(tmm[vargenes,]))
#pc_deseq <- fastpca(log1p(deseq))

colMerged <- col1[level1]
plot(pc_raw, col=colMerged, pch=20, main="PCA RAW")
legend("topright", levels(level1), fill=col1)

plot(pc_tc, col=colMerged, pch=20, main="PCA TC")
plot(pc_tmm, col=colMerged, pch=20, main="PCA TMM")
plot(pc_fq, col=colMerged, pch=20, main="PCA FQ")

colMerged <- col1[batch]
plot(pc_raw, col=colMerged, pch=20, main="PCA RAW")
legend("topright", levels(batch), fill=col1)

plot(pc_tc, col=colMerged, pch=20, main="PCA TC")
plot(pc_tmm, col=colMerged, pch=20, main="PCA TMM")
plot(pc_fq, col=colMerged, pch=20, main="PCA FQ")
```

## ZIFA

```{r zifa}
wrapRzifa <- function(Y, block = TRUE, k=2){
  # wrapper R function for ZIFA.
  # md5 hashing and temporary files are used not to re-run zifa 
  # if it has already be run on this computer.
  d = digest(Y, "md5")
  tmp = paste0(tempdir(), '/', d)
  write.csv(Y, paste0(tmp, '.csv'))
  
  if (!file.exists(paste0(tmp, "_", k, '_zifa.csv'))){
    print('run ZIFA')
    bb = ifelse(block, '-b ', '')
    cmd = sprintf('python run_zifa.py -d %d %s%s.csv %s_%d_zifa.csv', k, bb, tmp, tmp, k)
    system(cmd)
  }
  read.csv(sprintf("%s_%d_zifa.csv", tmp, k), header=FALSE)
}

zifa_raw <- wrapRzifa(log1p(raw[vargenes,]), k=2)
pairs(zifa_raw, pch=19, col=colMerged, main="ZIFA Raw")

zifa_tc <- wrapRzifa(log1p(tc[vargenes,]), k=2)
pairs(zifa_tc, pch=19, col=colMerged, main="ZIFA TC")

zifa_tmm <- wrapRzifa(log1p(tmm[vargenes,]), k=2)
pairs(zifa_tmm, pch=19, col=colMerged, main="ZIFA TMM")

zifa_fq <- wrapRzifa(log1p(fq[vargenes,]), k=2)
pairs(zifa_fq, pch=19, col=colMerged, main="ZIFA FQ")
```

## ZINB

```{r zinb}
raw <- as.matrix(raw)
zinb <- zinbFit(raw[vargenes,], K = 2, epsilon=1e3)
plot(zinb@W, pch=19, col=colMerged, main="ZINB")
```

## Add batches in ZINB model

```{r zinb_batch}
subset <- which(level2 %in% levels(level2)[c(1, 3, 4, 8)])

vars <- rowVars(log1p(tc[,subset]))
names(vars) <- rownames(tc)
vars <- sort(vars, decreasing = TRUE)
vargenes2 <- names(vars)[1:1000]

zinb_sub <- zinbFit(raw[vargenes2, subset], K = 2)
plot(getW(zinb_sub), pch=19, col=colMerged[subset], main="ZINB; color-coded by patient")
plot(getW(zinb_sub), pch=19, col=col2[droplevels(level2[subset])], main="ZINB; color-coded by inferred cell type")

zinb_batch <- zinbFit(raw[vargenes2, subset], K = 2, X=model.matrix(~batch[subset]))
plot(getW(zinb_batch), pch=19, col=colMerged[subset], main="ZINB w/ batch; color-coded by patient")
plot(getW(zinb_batch), pch=19, col=col2[droplevels(level2[subset])], main="ZINB w/ batch; color-coded by inferred cell type")

pc_fq_sub <- fastpca(log1p(fq[vargenes2, subset]))
plot(pc_fq_sub, pch=19, col=colMerged[subset], main="PCA (FQ); color-coded by patient")
plot(pc_fq_sub, pch=19, col=col2[droplevels(level2[subset])], main="PCA (FQ); color-coded by inferred cell type")
```

# Censoring model

```{r zinb_cdr}
Xmod <- model.matrix(~detection_rate)
zinb_cdr <- zinbFit(raw[vargenes,], K = 2, X = Xmod, which_X_mu = 1L, which_X_pi = 1:2, epsilon=1e3)

plot(zinb@W, pch=19, col=colMerged, main="ZINB")
plot(zinb_cdr@W, pch=19, col=colMerged, main="ZINB")

boxplot(detection_rate~batch)
```

# Combat

```{r combat}
library(sva)
combat_fq  <- ComBat(log1p(fq), batch)
pc_combat_fq  <- fastpca(combat_fq[vargenes,])
plot(pc_combat_fq, col=colMerged, pch=20, main="PCA FQ COMBAT")
```

## Considering only data for MGH26

```{r zinb_subset}
subset <- which(level1=="MGH26")

vars <- rowVars(log1p(tc[,subset]))
names(vars) <- rownames(tc)
vars <- sort(vars, decreasing = TRUE)
vargenes2 <- names(vars)[1:1000]

zinb_sub1 <- zinbFit(raw[vargenes2, subset], K=2)
plot(getW(zinb_sub1), pch=19, col=col1[droplevels(batch[subset])], main="ZINB; color-coded by batch")

zinb_sub2 <- zinbFit(raw[vargenes2, subset], K = 2, X=model.matrix(~batch[subset]))
plot(getW(zinb_sub2), pch=19, col=col1[droplevels(batch[subset])], main="ZINB w/batch correction")

zinb_sub3 <- zinbFit(raw[vargenes2, subset], K = 2, X=model.matrix(~detection_rate[subset]),
                     which_X_mu = 1L, which_X_pi = 1:2)
plot(getW(zinb_sub3), pch=19, col=col1[droplevels(batch[subset])], main="ZINB w/cdr covariate")
```

# Gene-level covariates

```{r gc_content, eval=FALSE}
counts <- counts[bm[,1],]
gene_info = getGeneLengthAndGCContent(as.character(bm[,2]), "mm10", mode="org.db")
rownames(gene_info) = bm[,1]
gene_info = na.omit(gene_info)
```

```{r write_out, dependson=c("zinb", "zifa", "pca", "zinb_batch")}
save(zinb, zinb_batch, zifa_fq, zifa_tmm, zifa_tc, zifa_raw, pc_tmm, pc_fq, pc_tc, pc_raw,
     zinb_cdr, pc_combat_fq, batch, 
     zinb_sub1, zinb_sub2, zinb_sub3, file="patel_covariates.rda")
```
