---
title: "Interpret clustering results"
author: "Davide Risso"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output: 
  html_document: 
    fig_height: 7
    fig_width: 7
    toc: yes
    code_folding: hide
    toc_float: yes
---

```{r options, include=FALSE}
knitr::opts_chunk$set(cache=TRUE, error=FALSE, message=FALSE, warning=FALSE)
```

In a previous document, we have completed a cluster analysis of the data after
dimensionality reduction with zinbwave (K=3).

We can see from the table below that the results are not the same that one gets
with the typical workflow proposed by TENx. Here we want to give a biological
interpretation to the clusters, to show that our clustering leads to more 
coherent biological groups.

```{r datain}
library(zinbwave)
library(cellrangerRkit)

tsne_coords <- read.csv("tenx_pbmc4k/projection.csv", row.names=1)
clusters <- read.csv("tenx_pbmc4k/clusters.csv")
cl <- clusters[,2]
names(cl) <- clusters[,1]
pal <- clusterExperiment::bigPalette

load("tenx_pbmc4k/zinb_res.rda")
load("tenx_pbmc4k/zinb_gcc.rda")

zcl <- read.table("tenx_pbmc4k/zinb_gcc_clusters.txt")
cl2 <- zcl[,4]

zcl <- read.table("tenx_pbmc4k/zinb_res_clusters.txt")
cl3 <- zcl[,4]

table(cl, cl2)
table(cl, cl3)
table(cl3, cl2)

pipestance_path <- "tenx_pbmc4k"
if(!file.exists(paste0(pipestance_path, "/outs"))) {
  download_sample(sample_name="pbmc4k",sample_dir=pipestance_path,
                  host="http://cf.10xgenomics.com/samples/cell-exp/1.3.0/")
}
pbmc <- load_cellranger_matrix(pipestance_path)
use_genes <- get_nonzero_genes(pbmc)
dense <- as.matrix(exprs(pbmc[use_genes, ]))

filter <- rowSums(dense>1)>=10
table(filter)
dense_all <- dense[filter,]

vars <- rowVars(log1p(dense))
names(vars) <- rownames(dense)
vars <- sort(vars, decreasing = TRUE)
vargenes <- names(vars)[1:1000]
dense <- dense[vargenes,]
```

To do so, we will first compute the normalized values using zinbwave and use 
them as an input to xCell which performs single-sample GSEA with cell type specific genesets.
We can compare the original 10x clusters, the clusters found by Seurat without any
normalization and the clusters found by Seurat after regressing out the total number
of UMIs and the percent of mitochondrial genes.

```{r norm}
se <- SummarizedExperiment(dense_all)
run_zinbwave <- FALSE

if(run_zinbwave) {
  library(BiocParallel)
  library(doParallel)
  registerDoParallel(6)
  register(DoparParam())
  system.time(zinbwave_res <- zinbwave(se, K = 3))
  save(zinbwave_res, file="tenx_pbmc4k/zinbwave_res.rda")
} else {
  load("tenx_pbmc4k/zinbwave_res.rda")
}

norm <- assay(zinbwave_res, "normalizedValues")
rownames(norm) <- fData(pbmc)[rownames(norm), 2]
```

```{r norm_plot}
plot(getW(zinb_gcc), pch=19, col=pal[cl2])
plot(reducedDim(zinbwave_res), pch=19, col=pal[cl2])
```

```{r xcell}
## manually write ssGSEA (workaround to circumvent error with xcell)
my.rawEnrichmentAnalysis <- function (expr, signatures, genes, file.name = NULL)  {
    shared.genes <- intersect(rownames(expr), genes)
    print(paste("Num. of genes:", length(shared.genes)))
    expr <- expr[shared.genes, ]
    # if (dim(expr)[1] < 5000) {
    #     print(paste("ERROR: not enough genes"))
    #     return - 1
    # }
    expr <- apply(expr, 2, rank)
    scores <- gsva(expr, signatures, method = "ssgsea", ssgsea.norm = FALSE)
    scores = scores - apply(scores, 1, min)
    cell_types <- unlist(strsplit(rownames(scores), "%"))
    cell_types <- cell_types[seq(1, length(cell_types), 3)]
    agg <- aggregate(scores ~ cell_types, FUN = mean)
    rownames(agg) <- agg[, 1]
    scores <- agg[, -1]
    if (!is.null(file.name)) {
        write.table(scores, file = file.name, sep = "\t", col.names = NA, 
            quote = FALSE)
    }
    scores
}
my.xCellAnalysis <- function (expr, signatures = NULL, genes = NULL, spill = NULL, 
    rnaseq = TRUE, file.name = NULL, scale = TRUE, alpha = 0.5, 
    save.raw = FALSE) 
{
    if (is.null(signatures)) 
        signatures = xCell.data$signatures
    if (is.null(genes)) 
        genes = xCell.data$genes
    if (is.null(spill)) {
        if (rnaseq == TRUE) {
            spill = xCell.data$spill
        }
        else {
            spill = xCell.data$spill.array
        }
    }
    if (is.null(file.name) || save.raw == FALSE) {
        fn <- NULL
    }
    else {
        fn <- paste0(file.name, "_RAW.txt")
    }
    scores <- my.rawEnrichmentAnalysis(expr, signatures, genes, 
        fn)
    scores.transformed <- transformScores(scores, spill$fv, scale)
    if (is.null(file.name)) {
        fn <- NULL
    }
    else {
        fn <- file.name
    }
    scores.adjusted <- spillOver(scores.transformed, spill$K, 
        alpha, fn)
    scores.adjusted = microenvironmentScores(scores.adjusted)
    return(scores.adjusted)
}

run_xcell <- FALSE

norm10x = function (ei)
{
  sums = colSums(ei)
  eo = t(t(ei)*median(sums)/sums)
  return(eo)
}

tc <- norm10x(as.matrix(exprs(pbmc[use_genes,])))
rownames(tc) <- fData(pbmc[use_genes])[,2]

if(run_xcell) {
  library(xCell)
  res_all <- xCellAnalysis(tc)
  save(res_all, file="tenx_pbmc4k/xcell_analysis_tc.rda")
} else {
  load("tenx_pbmc4k/xcell_analysis_tc.rda")
}
```

```{r xcell_res}
theme_set(theme_classic())

Wgcc <- getW(zinb_gcc)
df <- data.frame(W1=Wgcc[,1], W2=Wgcc[,2], W3=Wgcc[,3], Cluster=cl2, Cluster10X=cl, t(res_all))

ggplot(aes(W1, W2, colour=B.cells), data = df) + 
 geom_point() + scale_colour_gradient(low="blue", high="yellow")

ggplot(aes(W1, W2, colour=CD4..T.cells), data = df) + 
 geom_point() + scale_colour_gradient(low="blue", high="yellow")

ggplot(aes(W1, W2, colour=CD8..T.cells), data = df) + 
 geom_point() + scale_colour_gradient(low="blue", high="yellow")

ggplot(aes(W1, W2, colour=NK.cells), data = df) + 
 geom_point() + scale_colour_gradient(low="blue", high="yellow")

ggplot(aes(W1, W2, colour=Monocytes), data = df) + 
 geom_point() + scale_colour_gradient(low="blue", high="yellow")

ggplot(aes(W1, W2, colour=pDC), data = df) + 
 geom_point() + scale_colour_gradient(low="blue", high="yellow")

ggplot(aes(W1, W2, colour=cDC), data = df) + 
 geom_point() + scale_colour_gradient(low="blue", high="yellow")


sapply(1:max(cl), function(i) {
  names(which.max(rowMeans(res_all[,cl==i]) - rowMeans(res_all[,cl!=i])))
})

sapply(levels(cl2), function(i) {
  names(which.max(rowMeans(res_all[,cl2==i]) - rowMeans(res_all[,cl2!=i])))
})
```

The xCell method doesn't seem to work well. What we can do is a more classic approach
in which we look for differentially expressed genes per each cluster and look at
enrichment analysis of the DE genes. We can use the xCell signatures to look for enrichment
for each cell type.

# Differential expression

To do differential expression analysis, we will use the clusterExperiment package.
We will create an object with the normalized data and the cluster labels.
We will then look for DE genes using the findBestFeature() function.

## Our clustering

```{r clusterX}
library(clusterExperiment)
ce <- clusterExperiment(tc, clusters=cbind(as.character(cl2),
                                           paste0("tenx", cl),
                                           as.character(cl3)),
                        transformation = log1p)

colcl <- pal[1:9]
names(colcl) <- levels(cl2)
clusterLegend(ce)[[1]][,"color"] <- colcl[clusterLegend(ce)[[1]][,"name"]]

colcl <- pal[10:19]
names(colcl) <- paste0("tenx", 1:9)
clusterLegend(ce)[[2]][,"color"] <- colcl[clusterLegend(ce)[[2]][,"name"]]

colcl <- pal[20:29]
names(colcl) <- levels(cl3)
clusterLegend(ce)[[3]][,"color"] <- colcl[clusterLegend(ce)[[3]][,"name"]]

ce <- makeDendrogram(ce, dimReduce="PCA", ndims=10)

## find DE genes
top50 <- getBestFeatures(ce, contrastType = "OneAgainstAll", number=50)
top10 <- getBestFeatures(ce, contrastType = "OneAgainstAll", number=10)
head(top10)

plotHeatmap(ce, clusterSamplesData = "dendrogramValue", 
            clusterFeaturesData = unique(top10$IndexInOriginal),
            whichClusters = "all",
            main = "Top 10 genes per cluster")

plotHeatmap(ce, clusterSamplesData = "dendrogramValue", 
            clusterFeaturesData = unique(top50$IndexInOriginal),
            whichClusters = "all",
            main = "Top 50 genes per cluster")
```

## Their clustering

```{r clusterX2}
primaryClusterIndex(ce) <- 2
ce
ce <- makeDendrogram(ce, dimReduce="PCA", ndims=10)

## find DE genes
top50_tx <- getBestFeatures(ce, contrastType = "OneAgainstAll", number=50)
top10_tx <- getBestFeatures(ce, contrastType = "OneAgainstAll", number=10)
head(top10_tx)

plotHeatmap(ce, clusterSamplesData = "dendrogramValue", 
            clusterFeaturesData = unique(top10_tx$IndexInOriginal),
            whichClusters = "all",
            main = "Top 10 genes per cluster")

plotHeatmap(ce, clusterSamplesData = "dendrogramValue", 
            clusterFeaturesData = unique(top50_tx$IndexInOriginal),
            whichClusters = "all",
            main = "Top 50 genes per cluster")
```

## Our clustering without gc content

```{r clusterX3}
primaryClusterIndex(ce) <- 3
ce
ce <- makeDendrogram(ce, dimReduce="PCA", ndims=10)

## find DE genes
top50_nogc <- getBestFeatures(ce, contrastType = "OneAgainstAll", number=50)
top10_nogc <- getBestFeatures(ce, contrastType = "OneAgainstAll", number=10)
head(top10_nogc)

plotHeatmap(ce, clusterSamplesData = "dendrogramValue", 
            clusterFeaturesData = unique(top10_nogc$IndexInOriginal),
            whichClusters = "all",
            main = "Top 10 genes per cluster")

plotHeatmap(ce, clusterSamplesData = "dendrogramValue", 
            clusterFeaturesData = unique(top50_nogc$IndexInOriginal),
            whichClusters = "all",
            main = "Top 50 genes per cluster")
```

# Gene-set enrichment analysis

```{r romer}
library(limma)
library(xCell)
library(parallel)

## extract genesets from xcell
genesets <- unlist(geneIds(xCell.data$signatures))
celltypes <- sapply(strsplit(names(genesets), "%"), function(x) x[1])
names(genesets) <- NULL
gs <- tapply(genesets, celltypes, c)

gs_idx <- ids2indices(gs, rownames(ce), remove.empty=TRUE)

## create design matrix
primaryClusterIndex(ce) <- 1
ce
ce <- makeDendrogram(ce, dimReduce="PCA", ndims=10)
plotDendrogram(ce)

design <- model.matrix(~cl2 - 1)
cont <- clusterContrasts(ce, contrastType="Dendro")$contrastMatrix

## run GSEA
run_romer <- FALSE

if(run_romer) {
  romer_out <- mclapply(seq_len(NCOL(cont)), function(i) {
    romer(transform(ce), gs_idx, design, cont[,i])
  }, mc.cores = 6)
  save(romer_out, file="tenx_pbmc4k/romer_zinb.rda")
} else {
  load("tenx_pbmc4k/romer_zinb.rda")
}

lapply(romer_out, topRomer, alt="up")
lapply(romer_out, topRomer, alt="down")

## find DE genes
top_dendro <- getBestFeatures(ce, contrastType = "Dendro", number=50)
head(top_dendro)

plotHeatmap(ce, clusterSamplesData = "dendrogramValue", 
            clusterFeaturesData = unique(top_dendro$IndexInOriginal),
            whichClusters = "all",
            main = "Top Dendro genes")

Wgcc <- getW(zinb_gcc)
means <- sapply(gs, function(idx) colMeans(transform(ce[intersect(idx, rownames(ce)),])))
df <- data.frame(W1=Wgcc[,1], W2=Wgcc[,2], W3=Wgcc[,3], Cluster=cl2, Cluster10X=paste0("TenX", cl), means)

ggplot(aes(W1, W2, colour=Cluster), data = df) + 
 geom_point() + scale_color_manual(values = pal)

ggplot(aes(W1, W2, colour=Monocytes), data = df) + 
 geom_point() + scale_colour_gradient(low="blue", high="yellow")

ggplot(aes(W1, W2, colour=Keratinocytes), data = df) + 
 geom_point() + scale_colour_gradient(low="blue", high="yellow")

ggplot(aes(W1, W2, colour=DC), data = df) + 
 geom_point() + scale_colour_gradient(low="blue", high="yellow")

ggplot(aes(W1, W2, colour=aDC), data = df) + 
 geom_point() + scale_colour_gradient(low="blue", high="yellow")

ggplot(aes(W1, W2, colour=cDC), data = df) + 
 geom_point() + scale_colour_gradient(low="blue", high="yellow")

ggplot(aes(W1, W2, colour=pDC), data = df) + 
 geom_point() + scale_colour_gradient(low="blue", high="yellow")

ggplot(aes(W1, W2, colour=B.cells), data = df) + 
 geom_point() + scale_colour_gradient(low="blue", high="yellow")

ggplot(aes(W1, W2, colour=CD4..T.cells), data = df) + 
 geom_point() + scale_colour_gradient(low="blue", high="yellow")

ggplot(aes(W1, W2, colour=CD8..T.cells), data = df) + 
 geom_point() + scale_colour_gradient(low="blue", high="yellow")

ggplot(aes(W1, W2, colour=NK.cells), data = df) + 
 geom_point() + scale_colour_gradient(low="blue", high="yellow")
```

These results are not extremely easy to interpret but this is my best attempt:

- Cluster 0: Monocytes / Keratinocytes
- Cluster 1: T cells (?)
- Cluster 2: CD4+ T-cells?
- Cluster 3: CD8+ T-cells?
- Cluster 4: B-cells
- Cluster 5: NK cells?
- Cluster 6: aDC?
- Cluster 7: cDC
- Cluster 8: pDC or Plasma Cells or doublets?

## Boxplots of our clustering

```{r box1}
ggplot(aes(Cluster, Monocytes, fill=Cluster), data = df) +
  geom_boxplot() + ggtitle("Monocytes") + scale_fill_manual(values = pal)

ggplot(aes(Cluster, Keratinocytes, fill=Cluster), data = df) +
  geom_boxplot() + ggtitle("Keratinocytes") + scale_fill_manual(values = pal)

ggplot(aes(Cluster, DC, fill=Cluster), data = df) +
  geom_boxplot() + ggtitle("DC") + scale_fill_manual(values = pal)

ggplot(aes(Cluster, cDC, fill=Cluster), data = df) +
  geom_boxplot() + ggtitle("cDC") + scale_fill_manual(values = pal)

ggplot(aes(Cluster, pDC, fill=Cluster), data = df) +
  geom_boxplot() + ggtitle("pDC") + scale_fill_manual(values = pal)

ggplot(aes(Cluster, B.cells, fill=Cluster), data = df) +
  geom_boxplot() + ggtitle("B cells") + scale_fill_manual(values = pal)

ggplot(aes(Cluster, CD4..T.cells, fill=Cluster), data = df) +
  geom_boxplot() + ggtitle("CD4+ T-cells") + scale_fill_manual(values = pal)

ggplot(aes(Cluster, CD8..T.cells, fill=Cluster), data = df) +
  geom_boxplot() + ggtitle("CD8+ T-cells") + scale_fill_manual(values = pal)

ggplot(aes(Cluster, NK.cells, fill=Cluster), data = df) +
  geom_boxplot() + ggtitle("NK cells") + scale_fill_manual(values = pal)
```

## Boxplots of 10X clustering

```{r box2}
ggplot(aes(Cluster10X, Monocytes, fill=Cluster10X), data = df) +
  geom_boxplot() + ggtitle("Monocytes") + scale_fill_manual(values = pal[10:19])

ggplot(aes(Cluster10X, Keratinocytes, fill=Cluster10X), data = df) +
  geom_boxplot() + ggtitle("Keratinocytes") + scale_fill_manual(values = pal[10:19])

ggplot(aes(Cluster10X, DC, fill=Cluster10X), data = df) +
  geom_boxplot() + ggtitle("DC") + scale_fill_manual(values = pal[10:19])

ggplot(aes(Cluster10X, cDC, fill=Cluster10X), data = df) +
  geom_boxplot() + ggtitle("cDC") + scale_fill_manual(values = pal[10:19])

ggplot(aes(Cluster10X, pDC, fill=Cluster10X), data = df) +
  geom_boxplot() + ggtitle("pDC") + scale_fill_manual(values = pal[10:19])

ggplot(aes(Cluster10X, B.cells, fill=Cluster10X), data = df) +
  geom_boxplot() + ggtitle("B cells") + scale_fill_manual(values = pal[10:19])

ggplot(aes(Cluster10X, CD4..T.cells, fill=Cluster10X), data = df) +
  geom_boxplot() + ggtitle("CD4+ T-cells") + scale_fill_manual(values = pal[10:19])

ggplot(aes(Cluster10X, CD8..T.cells, fill=Cluster10X), data = df) +
  geom_boxplot() + ggtitle("CD8+ T-cells") + scale_fill_manual(values = pal[10:19])

ggplot(aes(Cluster10X, NK.cells, fill=Cluster10X), data = df) +
  geom_boxplot() + ggtitle("NK cells") + scale_fill_manual(values = pal[10:19])
```

# Session Info

```{r}
sessionInfo()
```
